dev-tools/mage/version/version_parser.go

```diff
@@ -0,0 +1,190 @@
+const preReleaseSeparator = "-"
+const metadataSeparator = "+"
+
+var semVerFmtRegEx *regexp.Regexp
+var namedGroups map[string]int
+
+func init() {
+	// small init to compile the regex and build a map of named groups and indexes
+	semVerFmtRegEx = regexp.MustCompile(semVerFormat)
+	groups := semVerFmtRegEx.SubexpNames()
+	namedGroups = make(map[string]int, len(groups))
+	for i, groupName := range groups {
+		namedGroups[groupName] = i
+	}
+}
+
+var ErrNoMatch = errors.New("version string does not match expected format")
+
+type ParsedSemVer struct {
+	original      string
+	major         int
+	minor         int
+	patch         int
+	prerelease    string
+	buildMetadata string
+}
+
+func (psv ParsedSemVer) Original() string {
+	return psv.original
+}
+
+func (psv ParsedSemVer) Major() int {
+	return psv.major
+}
+
+func (psv ParsedSemVer) Minor() int {
+	return psv.minor
+}
+
+func (psv ParsedSemVer) Patch() int {
+	return psv.patch
+}
+
+func (psv ParsedSemVer) CoreVersion() string {
+	return fmt.Sprintf("%d.%d.%d", psv.Major(), psv.Minor(), psv.Patch())
+}
+
+func (psv ParsedSemVer) Prerelease() string {
+	return psv.prerelease
+}
+
+func (psv ParsedSemVer) BuildMetadata() string {
+	return psv.buildMetadata
+}
+
+func (psv ParsedSemVer) VersionWithPrerelease() string {
+	b := new(strings.Builder)
+	b.WriteString(psv.CoreVersion())
+	if psv.prerelease != "" {
+		b.WriteString("-")
+		b.WriteString(psv.prerelease)
+	}
+	return b.String()
+}
+
+func (psv ParsedSemVer) IsSnapshot() bool {
+	return psv.prerelease == "SNAPSHOT" || strings.HasSuffix(psv.prerelease, "-SNAPSHOT")
+}
+
+func (psv ParsedSemVer) Less(other ParsedSemVer) bool {
+	// compare major version
+	if psv.major != other.major {
+		return psv.major < other.major
+	}
+
+	//same major, check minor
+	if psv.minor != other.minor {
+		return psv.minor < other.minor
+	}
+
+	//same minor, check patch
+	if psv.patch != other.patch {
+		return psv.patch < other.patch
+	}
+
+	// last resort check if one is prereleas and the other isn't
+	if psv.prerelease != "" && other.prerelease == "" {
+		return true
+	}
+
+	return false
+}
+
+func (psv ParsedSemVer) GetPreviousMinor() (*ParsedSemVer, error) {
+	major := psv.Major()
+	minor := psv.Minor()
+
+	if minor > 0 {
+		// We have at least one previous minor version in the current
+		// major version series. Set the patch to zero to guarnatee the
+		// version exists, the number of patch releases varies.
+		return NewParsedSemVer(major, minor-1, 0, psv.Prerelease(), psv.BuildMetadata()), nil
+	}
+
+	// We are at the first minor of the current major version series. To
+	// figure out the previous minor, we need to rely on knowledge of
+	// the release versions from the past major series'.
+	switch major {
+	case 8:
+		return NewParsedSemVer(7, 17, 10, psv.Prerelease(), psv.BuildMetadata()), nil
+	}
```

Comment:
A switch statement with one case is just an if statement:
```suggestion
	if major == 8 {
		return NewParsedSemVer(7, 17, 10, psv.Prerelease(), psv.BuildMetadata()), nil
	}
```